great_expectations/rule_based_profiler/domain_builder/categorical_column_domain_builder.py

```diff
@@ -0,0 +1,247 @@
+
+        Returns:
+            List of dicts of the form [{column_name: List[MetricConfiguration]},...]
+        """
+
+        limit_mode: CardinalityLimitMode = self._cardinality_checker.limit_mode
+
+        metric_name: str = limit_mode.related_metric_name
+
+        metric_configurations: List[Dict[str, List[MetricConfiguration]]] = []
+
+        for column_name in column_names:
+            metric_configurations.append(
+                {
+                    column_name: [
+                        MetricConfiguration(
+                            metric_name=metric_name,
+                            metric_domain_kwargs={
+                                "batch_id": batch_id,
+                                "column": column_name,
+                            },
+                            metric_value_kwargs=None,
+                            metric_dependencies=None,
+                        )
+                        for batch_id in batch_ids
+                    ]
+                }
+            )
+
+        return metric_configurations
+
+    def _columns_meeting_cardinality_limit(
+        self,
+        validator: Validator,
+        table_column_names: List[str],
+        metrics_for_cardinality_check: List[Dict[str, List[MetricConfiguration]]],
+    ) -> List[str]:
+        """Compute cardinality and return column names meeting cardinality limit.
+
+        Args:
+            validator: Validator used to compute column cardinality.
+            table_column_names: column names to verify cardinality.
+            metrics_for_cardinality_check: metric configurations used to
+                compute cardinality.
+
+        Returns:
+            List of column names meeting cardinality.
+        """
+        candidate_column_names: List[str] = table_column_names.copy()
+
+        # TODO AJB 20220218: Remove loops and refactor into single call to validator.get_metrics()
+        #  by first building up MetricConfigurations to compute. See GREAT-584
+        for metric_for_cardinality_check in metrics_for_cardinality_check:
+            for (
+                column_name,
+                metric_config_batch_list,
+            ) in metric_for_cardinality_check.items():
+                if column_name not in candidate_column_names:
+                    continue
+                else:
+                    for metric_config in metric_config_batch_list:
+
+                        metric_value = validator.get_metric(metric=metric_config)
+
+                        batch_cardinality_within_limit: bool = (
+                            self._cardinality_checker.cardinality_within_limit(
+                                metric_value=metric_value
+                            )
+                        )
+
+                        if batch_cardinality_within_limit is False:
+                            candidate_column_names.remove(column_name)
+                            break
```

Comment:
@anthonyburdi Why do we `break` out of the loop -- might there be other columns to filter out?..  Also, please see an earlier comment about building columns list, instead of starting the the maximum possible list and then filtering it out.  Thanks!

Reply:
We are breaking out of the inner loop, so we stop considering that column since it fails the cardinality test. Then we continue with the other columns in the outer loop. I hope soon we can use `validator.get_metrics()` to compute these all at once. It will depend on the data / data store whether one method is faster than another.

Reply:
@anthonyburdi So it cannot un-fail for the next batch?

Reply:
@anthonyburdi Please wait for #4282 to get merged, if you can -- you get parallel execution!  Thanks!

Reply:
@alexsherstinsky correct - it can't un-fail for the next batch